src/chat/chat_panel.rs

```diff
@@ -929,6 +869,86 @@ impl ChatPanel {
             auto_scroll_cancellable: false,
         };
     }
+
+    fn update_visibilities(&mut self, scope: &mut Scope) {
```

Comment:
Keeping visibility controls in a single place makes the code much more readable and made me easy to debug all the issues :)

Although, this is just a small improvement. Ideally I believe view logic and event handling logic should be decoupled even further.

Reply:
even though the function itself makes sense, i believe splitting the state is a mistake here, for a couple of reasons: first is that having two data structures without a real difference in what they represent is worse that centralizing them in one, they are both the state of the widget, why separating them? second is that you've flattened the state here, its not clear when the two enums exists at each point of the state of the widget, adding more cognitive load and problems in the future. by keeping it in the same one we could draw a graph with the state and its transitions more easily, making it easier for the widget to act over it now and better for maintainability.

Reply:
> two data structures without a real difference
As I mentioned in a comment, `ComputedState` is not necessary. It's just for organization. I can remove it if you prefer. But it's not the same as what is called "state" in the widget.

>  is worse that centralizing them in one,

Yes, this centralizes state. The state is centralized at the store. And is computed from there, instead of duplicating it (which is what `ChatPanelState` was doing). The `enum` is just for code organization and can be removed by putting the code directly in the corresponding if blocks (but that's less readable in my opinion). Maybe the world `state` is confusing you? 

`ChatPanelState` stores stuff that already exists in the store and can be computed from there, duplicating data unnecessary and provoking two sources of truths. 

Maybe with a little more effort I can get rid of `ChatPanelState` completely (instead of partially as I did) but that's out of the scope of this task.

Reply:
https://github.com/moxin-org/moxin/commit/425f579a638512ee419e9db812478d3eaf3189ce this is how it would look without the helper enum. Probably looking it like this helps to resolve the confusion. This is 100% equivalent. There is absolutely no difference (just that it has some stuff duplicated, a lot of long nesting and is arguably less readable).

Reply:
the point of ChatPanelState is to have a representation of the state of the widget, its not "data" as in the store. we can make the widget act over it, and make the transitions based on the store or whatever we want separately, without it we would be making boolean checks all over the place, like the ones you're doing in this function, or keeping single boolean variables. right now you're mixing both approaches, which is even worse.

Reply:
> the point of ChatPanelState is to have a representation of the state of the widget

But this is currently "app state", no "local state". Like is not state of the widget itself, but state from somewhere that the widget uses to compute it's view. 

100% proof of that this is no local state is that you can compute it completely. You could have a `state()` function that computes the widget state from the `store` returning some kind of `State` and that would work without storing it in the widget struct.

The widget struct  doesn't need to hold anything. Holding it may act as a kind of "cache" if computations are expensive, but that would be premature optimization.

> we can make the widget act over it, and make the transitions based on the store or whatever we want separately, without it we would be making boolean checks all over the place, like the ones you're doing in this function

Why do you want to act on two things instead of one? Like if you need to send something like `load_model` to the store, and at the same time transition the local state to `Idle`, it may happen that `load_model` fails and the store would not transition to the next state (thinking it like finite state machines) but you updated your local state and now you have 2 source of truth saying two different things.

And this is not an hypotetical case, look at what is in main for an example:

```
    fn load_model(&mut self, store: &mut Store, downloaded_file: DownloadedFile) {
        self.state = ChatPanelState::Idle;
        self.view(id!(main)).set_visible(true);
        self.view(id!(empty_conversation)).set_visible(true);
        self.view(id!(no_model)).set_visible(false);
        self.view(id!(no_downloaded_model)).set_visible(false);

        store.load_model(&downloaded_file.file);
    }
```

Here you transition the local state to `Idle`, but the `load_model` that comes after it can fail. If it fails, the store doesn't transition but your local state did, generating an inconsitency that would need help of god to debug. 

In general, keeping two structures in sync is highly error-prone no matter the kind of app or framework you are using and it can be a maintainability madness. If one "state" can be computed from other "state", it should be.

Of course, if the widget actually had some kind of owned state, then it would be ok to have it in the widget, but this widget doesn't.

So, I must insist, keeping track of this in the widget is a bad idea, and can be the source of worst state issues eventually.